File: Questions 401 - 500/494_targetSum.py
def findTargetSumWays(nums, target):
    # maintain a cache
    # cache will store (index, total so far) -> number of ways
    dp = {}

    index = 0
    total = 0
    res = DFS(nums, index, total, dp, target)
    return res


def DFS(nums, index, total, dp, target):

    # 1 way if I have reached the last index and achieved target
    if(index == len(nums)):
        return 1 if total == target else 0

    if (index, total) in dp:
        return dp[(index, total)]

    # update cache with number of ways with addition and with subtraction
    dp[(index, total)] = DFS(nums, index + 1, total + nums[index], dp,
                             target) + DFS(nums, index + 1, total - nums[index], dp, target)

    return dp[(index, total)]


if __name__ == "__main__":

    nums = [1, 2, 3, 0, 4, -4]
    target = 0

    print(
        f"Number of ways to obtain target from nums is: {findTargetSumWays(nums, target)}")

File: Questions 401 - 500/494_targetSum.cpp
#include<iostream>
#include<vector>
#include<map>

using namespace std;

int DFS(vector<int> &nums, int index, int target, int total, map<pair<int, int>, int> &dp){
    
    if(index == nums.size()) {
        return (total == target) ? 1 : 0;
    } 
    
    if(dp.find({index, total}) != dp.end()){
        return dp[{index, total}];
    }
    
    else {
        dp[{index, total}] = DFS(nums, index + 1, target, total + nums[index], dp) 
                            + DFS(nums, index + 1, target, total - nums[index], dp);
    }
    
    return dp[{index, total}];
}


int findTargetSumWays(vector<int>& nums, int target) {
    map<pair<int,int>, int> dp;

    int index = 0;
    int total = 0;
    
    int res = DFS(nums, index, target, total, dp);
    
    return res;
}
    
int main() {

    vector<int> nums = {1,1,1,1,1};
    int target = 3;
    
    cout<<"Number of ways to obtain target: "<<findTargetSumWays(nums, target);

    return 0;
}